packages/connection-storage/src/connection-provider.ts

```diff
@@ -0,0 +1,90 @@
+import { ConnectionInfo } from '@mongodb-js/connection-info';
+import { ConnectionStorage } from './connection-storage';
+import ConnectionString from 'mongodb-connection-string-url';
+import { merge } from 'lodash';
+
+export type ConnectionStatus =
+  | 'connected'
+  | 'disconnected'
+  | 'connecting'
+  | 'failed';
+export type StatusAwareConnectionInfo = Omit<
+  ConnectionInfo & {
+    status: ConnectionStatus;
+  },
+  'favorite'
+>;
+
+/**
+ * Exposes an interface that will be implemented for each deployment environment to
+ * get access to the connections on that environment. For Compass Desktop, it will
+ * rely on the ConnectionStorage, for the Data Explorer, on the Atlas API (or another
+ * facade on top of that).
+ */
+export interface ConnectionProvider {
+  listConnections(): Promise<StatusAwareConnectionInfo[]>;
+  listConnectionHistory?(): Promise<StatusAwareConnectionInfo[]>;
+  saveConnection?(info: ConnectionInfo): Promise<StatusAwareConnectionInfo>;
+  deleteConnection?(info: ConnectionInfo): Promise<void>;
+}
+
+export class DesktopConnectionProvider implements ConnectionProvider {
```

Comment:
"Desktop" vs "Web": I personally really like this way to distinguish the implementations, though i'm not sure this is used consistently across the code base, can we double check and align on one convention?

Reply:
We typically use `Compass` as a name prefix for the desktop app since we're still referring to DE as a separate product

Reply:
In the TD it's specified as a tentative name and I'm thinking on proposing it as a general term on Monday 😄 . We already use `Compass Web` so `Compass Desktop` makes sense (instead of using electron or any other naming).